unic/utils/src/tables.rs

```diff
@@ -1,56 +1,89 @@
-//! Trait and impls for character data tables used in UNIC.
+//! Character data tables used in UNIC.
 
-use std::cmp;
+use unic_char_range::CharRange;
 
-/// A table from characters to data associated with those characters.
-pub trait CharDataTable<V> {
-    /// Find the associated data for a character in this table.
-    fn find(&self, needle: char) -> Option<V>;
+/// A mapping from characters to some associated data.
+///
+/// For the set case, use `()` as the associated value.
+#[derive(Copy, Clone, Debug)]
+pub enum CharDataTable<V: 'static> {
+    #[doc(hidden)] Direct(&'static [(char, V)]),
+    #[doc(hidden)] Range(&'static [(CharRange, V)]),
+}
 
-    /// Find the associated data for a character in this table or use a provided default.
-    fn find_or(&self, needle: char, default: V) -> V {
-        self.find(needle).unwrap_or(default)
+impl<V> Default for CharDataTable<V> {
+    fn default() -> Self {
+        CharDataTable::Direct(&[])
     }
+}
 
-    /// Find the associated data for a character in this table or generate a default.
-    fn find_or_else<F>(&self, needle: char, f: F) -> V
```

Comment:
Just wondering: How come decided to drop `find_or_else` and `find_or` and use a custom name?

Reply:
`find_or_else`/`find_or` are duplication of behavior from the `Option`. I think it's clearer to use the `Option` fn. That, and I'm not even really sure as to what the order of the parameters should be.

`find_defaulting` doesn't have any other call that clearly explains what's going on, and is a useful alias for `find_or_else(ch, Default::default)` (or `find(ch).unwrap_or_else(Default::default)` if we don't have `find_or_else`). That, and since I've bound it on `Copy` it makes working with these `Copy + Default` types a lot easier. The name might be better off as `find_or_default` though.

TL;DR: `find_or_else`/`find_or` don't add functionality over using `Option` fn, `find_or_default` (currently called `find_defaulting`) does.

Reply:
That, and `find_or_else`/`find_or` (the concepts) aren't used in any of the code since all of our tables are called using `find_defaulting`.

Reply:
Yeah, sounds good. Agree that `find_or_default()` would be a better name.